udf.go

```diff
@@ -0,0 +1,259 @@
+package duckdb
+
+/*
+#include <stdlib.h>
+#include <duckdb.h>
+
+void udf_bind(duckdb_bind_info info);
+
+void udf_init(duckdb_init_info info);
+
+void udf_callback(duckdb_function_info, duckdb_data_chunk);  // https://golang.org/issue/19837
+
+void udf_destroy_data(void *);
+
+typedef void (*init)(duckdb_function_info);  // https://golang.org/issue/19835
+typedef void (*bind)(duckdb_function_info);  // https://golang.org/issue/19835
+typedef void (*callback)(duckdb_function_info, duckdb_data_chunk);  // https://golang.org/issue/19835
+*/
+import "C"
+
+import (
+	"database/sql"
+	"reflect"
+	"runtime/cgo"
+	"unsafe"
+)
+
+type (
+	ColumnMetaData struct {
+		Name string
+		T    Type
+	}
+
+	CardinalityData struct {
+		Cardinality uint
+		IsExact     bool
+	}
+
+	TableFunctionInitData struct {
+		MaxThreads int
+	}
```

Comment:
Could you elaborate a bit more?

While thinking about this I did find a couple cases where I can see the benefit of adding localstate, even if it complicated the API more. But I could not come up with a scenario where it matters how many threads are running.

Reply:
Yes I agree that a concurrency-aware `FillRow` would be needed, and that maybe adding locks to `FillRow` would be detrimental to the performance, that is indeed true. However if I add a localstate a lot of locking could be worked around, and functions may be able to use atomics instead of locks to avoid synchronization overhead.

I might implement a vector based API sometime this weekend, maybe that will help with this.

The vector vs row discussion is indeed a bigger than just this PR, which is important as users want more performance (by using vectorised APIs). I started with a row based API because it is probably conceptually the simplest for users.

Reply:
I added threaded variants of the options, where the threading is exposed and a functions to create local state is required. This removes any worry of threading for those that do not need it, but exposes the full API when someone wants to use it.